### PR TITLE
core/consensus: add support for pointer values on the wire

### PR DESCRIPTION
**Pull request by corverroos**
Upgrades the QBFT wire protocol (without breaking backwards compatibility) by adding the whole value to the `ConsensusMsg` envelope and hashes of those values to the `QBFTMsg` itself. This will allow us to remove the values from the QBFT messages, which will reduce network bandwidth usage by a lot (90% for large values; beacon blocks).

This increases the network bandwidth for next release v0.14, but once we remove support for v0.13, it will reduce it by a lot.

category: feature
ticket: #1552

### CODE REVIEW

File: core/consensus/msg_internal_test.go [hunk 1]
```diff
@@ -73,150 +69,28 @@ func TestSigning(t *testing.T) {
 	require.False(t, ok)
 }
 
-func TestBackwardsCompatibility(t *testing.T) {
```
Comment (corverroos) [inline core/consensus/msg_internal_test.go:76]: this tests 2 updates ago, we already removed support, can remove now.

File: core/corepb/v1/consensus.proto [hunk 2]
```diff
@@ -13,31 +13,20 @@ message QBFTMsg {
...
+  bytes                          value_hash            = 11;
+  bytes                          prepared_value_hash   = 12;
 }
 
 message ConsensusMsg {
-  QBFTMsg          msg           = 1;
-  repeated QBFTMsg justification = 2;
+  QBFTMsg                      msg           = 1;
+  repeated QBFTMsg             justification = 2;
+  repeated google.protobuf.Any values        = 3;
```
Comment (corverroos) [inline core/corepb/v1/consensus.proto:29]: so the values are added here and can therefore be removed from `QBFTMsg` above once we remove support for v0.13.

File: core/consensus/component.go [hunk 3]
```diff
@@ -545,3 +568,22 @@ func newRoundTimer(round int64) (<-chan time.Time, func()) {
 	timer := time.NewTimer(roundStart + (time.Duration(round) * roundIncrease))
 	return timer.C, func() { timer.Stop() }
 }
+
+func valuesByHash(values []*anypb.Any) (map[[32]byte]*anypb.Any, error) {
+	resp := make(map[[32]byte]*anypb.Any)
```
Comment (gsora) [inline core/consensus/component.go:573]: Should this function return error if values is empty?

File: core/corepb/v1/consensus.proto (hunk 2 above)
Comment (gsora) [reply core/corepb/v1/consensus.proto:29]: Nit: can we add a comment for future reference?

**Review by gsora — APPROVED** [review]
General approve because it looks good.

I wonder if we could get rid of the `*anypb.Any` and deal with copies instead.

File: core/consensus/component.go (hunk 3 above)
Comment (corverroos) [reply core/consensus/component.go:573]: nope, empty values result in a empty map, this is required for legacy v0.13 support.

File: core/corepb/v1/consensus.proto (hunk 2 above)
Comment (corverroos) [reply core/corepb/v1/consensus.proto:29]: sure

**Comment by corverroos** [comment]
Since the `Anys` wrap protobufs, the wrapped objects are still pointers unfortunately.

**Pull request merged**